### ISSUE TITLE
WhatsApp "Someone" nicknaming

### ISSUE THREAD

**Issue by DummyThatMatters** [42wim/matterbridge#1363]
**Is your feature request related to a problem? Please describe.**
WhatsApp tend to relay messages with "\<Someone\>" nickname in some cases (randomly) even if actual WhatsApp nickname presented. So it is completely impossible to distinguish authors of the posts on the other side of the bridge.

**Describe the solution you'd like**
To fix WhatsApp "\<Someone\>" nicknames with actual nicknames and add the option to use phone numbers as \<NICK\> source (if actual nickname is not set).

**Describe alternatives you've considered**
To generate unique random pseudonym nickname for each "anonymous" user instead of static "\<Someone\>" nickname.

**Additional context**
Group chat, WhatsApp - Mattermost bridge, ~50 users, some users have specified nickname, some of them are not.

**Comment by JEK58** [comment]
Why did you close this? Did you finde a solution?

**Comment by joko2020** [comment]
Dear Sir,
I need help to integrate Whatsapp to Nextcloud Talk, I have Installed matterbridge on Nextcloud 20.0.7 with matterbridge Running version 1.22.0... but I cannot find whatsapp channel in the talk matterbridge setting. after googling seems to  manually edit the matterbridge.toml config file.. but i' cannot find that file.. where is the file path of matterbridge.toml  
thanks

**Comment by 42wim** [comment]
@joko2020 you should look at the nextcloud forums, I don't know where they put the matterbridge.toml

**Comment by joko2020** [comment]
> @joko2020 you should look at the nextcloud forums, I don't know where they put the matterbridge.toml

Thanks for your reply.. unfortunately , I cannot get the right inormation Also

**Comment by micw** [comment]
Solution for this particular issue is IMO to allow to show the phone number instead of "someone". While that's not wanted in public groups, it's absolutely wanted in most private group bridging.
I'll create another issue for the fact that jid->account nick name resolution seems to be broken.

**Comment by micw** [comment]
When reading the code og go-whatsapp, I found:
https://github.com/42wim/matterbridge/blob/4fcad8e04befeec2c7e468b36500fa33eb22af51/vendor/github.com/Rhymen/go-whatsapp/store.go#L48

May it be that the key in the contatc list is different from the actual sender jid, causing the lookup failures?

**Comment by micw** [comment]
It seems that if I restart matterbridge, after someone has sent a message to a WA group, this one's nick is resolved properly.

**Comment by micw** [comment]
There was already a related fix in https://github.com/42wim/matterbridge/issues/796. But it seems that this is insufficient. I have tried to analyze when `someone` is printed and when the nick name.
It seems that if a user posts a second message after he posts a message, the nick is resolved properly. After restart, the nick name cache seems to be re-filled but only from the recent messages. So if one posts now again and he had posted recently, the nick is resolved at the first message.

My assumption: When a message arrives and the nick is not known, it is retrieved somehow. But this happens _after_ the nick for this first message is already resolved to `someone`. But it should happen before this.

Could anyone with knowledge about the code verify (and hopefully fix) this?

**Comment by micw** [comment]
@nathanaelhoun I ping you in this ticket since your PR #1160 seems to be related to this issue.

**Comment by nathanaelhoun** [comment]
Hi @micw, I tried to fix this issue as you noted.
> It seems that if a user posts a second message after he posts a message, the nick is resolved properly. After restart, the nick name cache seems to be re-filled but only from the recent messages. So if one posts now again and he had posted recently, the nick is resolved at the first message.

I arrived to the same conclusion and it's why I made #1160 to reload the user list if their is no nickname. Because this changes did not solve the issue (at least, not everytime), I believe your assumption may be right :
> My assumption: When a message arrives and the nick is not known, it is retrieved somehow. But this happens after the nick for this first message is already resolved to someone. But it should happen before this

**Comment by micw** [comment]
@nathanaelhoun 
I'm reviewing
https://github.com/42wim/matterbridge/blob/78f3820cede34808fec7e90fe42e564379dd4546/bridge/whatsapp/helpers.go#L75-L111

To avoid code duplication (which leads to diverge in code as already happened, compare L77-L89 with L102-107) I think the check should be earlier, something like:
```
sender, exists := b.users[senderJid]
if  !exists {
  // reload, add, set exists=true
}
if  exists {
  // code from L77-L89
}
```

Besides that, `Contacts()` looks to me like is jkust sends out the request. The response will probably asynchronous. So message processing should somehow block until the contacts are available. This might be quite complicated because another message might arrive meanwhile which also triggers a contacts reload.

**Comment by gabuch2** [comment]
Is there a solution or workaround to this?
